src/renderers/WebGLRenderer.js

```diff
@@ -1950,12 +1875,6 @@ function WebGLRenderer( parameters ) {
 
 			WebGLUniforms.upload( _gl, materialProperties.uniformsList, m_uniforms, textures );
 
-			if ( material.isShaderMaterial ) {
-
-				material.uniformsNeedUpdate = false; // #15581
-
-			}
-
```

Comment:
Hmm, you changed the order of this code. Before the PR it executed after `WebGLUniforms.upload()` and now it's executing before.

I'll merge the PR but I'll put this code back just in case. Step by step...

Reply:
AFAICS, that should not have any side effects. It's just important that this bit comes before this section:
```js
if ( material.isShaderMaterial && material.uniformsNeedUpdate === true ) {

    WebGLUniforms.upload( _gl, materialProperties.uniformsList, m_uniforms, textures );
    material.uniformsNeedUpdate = false;

}
```
Otherwise `WebGLUniforms.upload()` is called twice for `ShaderMaterials` when `uniformsNeedUpdate` is set to `true`.

Reply:
ok. This makes me wonder, how much you guys trust unit and e2e tests?
And how safe is more radical refactoring?

Reply:
> This makes me wonder, how much you guys trust unit and e2e tests?

Right now, the test coverage is not sufficient so broad refactoring can be done without doing a lot of manual testing.

Reply:
> ok. This makes me wonder, how much you guys trust unit and e2e tests?

Unit tests and e2e are nice to have but they wouldn't catch uploading uniforms twice.

The codebase is still fairly small just so we can still keep it in our heads (more or less). I would hate if we get to a point where no one understands it and we just rely on tests.

Generally not a fan of radical refactorings. Incremental refactoring are more manageable.

Reply:
Okay, moved the code back: https://github.com/mrdoob/three.js/commit/b063139c82164382aa5572b0aa7e3f3e91cc06d2